plugins/outputs/kinesis/kinesis_test.go

```diff
@@ -83,3 +89,219 @@ func TestPartitionKey(t *testing.T) {
+
+	svc.AssertRequests(assert, []*kinesis.PutRecordsInput{
+		{
+			StreamName: &streamName,
+			Records:    records,
+		},
+	})
+}
+
+func TestWriteKinesis_WhenRecordErrors(t *testing.T) {
+
+	assert := assert.New(t)
+
+	errorCode := "InternalFailure"
+	errorMessage := "Internal Service Failure"
+	partitionKey := "partitionKey"
+	streamName := "stream"
+
+	records := []*kinesis.PutRecordsRequestEntry{
+		{
+			PartitionKey: &partitionKey,
+			Data:         []byte{0x66},
+		},
+	}
+
+	svc := &mockKinesisPutRecords{}
+	svc.SetupResponse(
+		1,
+		[]*kinesis.PutRecordsResultEntry{
+			{
+				ErrorCode:      &errorCode,
+				ErrorMessage:   &errorMessage,
+				SequenceNumber: nil,
+				ShardId:        nil,
+			},
+		},
+	)
+
+	k := KinesisOutput{StreamName: streamName}
+	k.mockKinesisService(svc)
+
+	elapsed := k.writeKinesis(records)
+	assert.Greater(elapsed.Nanoseconds(), zero)
+
+	svc.AssertRequests(assert, []*kinesis.PutRecordsInput{
+		{
+			StreamName: &streamName,
+			Records:    records,
+		},
+	})
+}
+
+func TestWriteKinesis_WhenServiceError(t *testing.T) {
+
+	assert := assert.New(t)
+
+	partitionKey := "partitionKey"
+	streamName := "stream"
+
+	records := []*kinesis.PutRecordsRequestEntry{
+		{
+			PartitionKey: &partitionKey,
+			Data:         []byte{},
+		},
+	}
+
+	svc := &mockKinesisPutRecords{}
+	svc.SetupErrorResponse(
+		awserr.New("InvalidArgumentException", "Invalid record", nil),
+	)
+
+	k := KinesisOutput{StreamName: streamName}
+	k.mockKinesisService(svc)
+
+	elapsed := k.writeKinesis(records)
+	assert.Greater(elapsed.Nanoseconds(), zero)
+
+	svc.AssertRequests(assert, []*kinesis.PutRecordsInput{
+		{
+			StreamName: &streamName,
+			Records:    records,
+		},
+	})
+}
+
+type mockKinesisPutRecordsResponse struct {
+	Output *kinesis.PutRecordsOutput
+	Err    error
+}
+
+type mockKinesisPutRecords struct {
+	kinesisiface.KinesisAPI
+
+	requests  []*kinesis.PutRecordsInput
+	responses []*mockKinesisPutRecordsResponse
+}
+
+func (m *mockKinesisPutRecords) SetupResponse(
+	failedRecordCount int64,
+	records []*kinesis.PutRecordsResultEntry,
+) {
+
+	m.responses = append(m.responses, &mockKinesisPutRecordsResponse{
+		Err: nil,
+		Output: &kinesis.PutRecordsOutput{
+			FailedRecordCount: &failedRecordCount,
+			Records:           records,
+		},
+	})
+}
+
+func (m *mockKinesisPutRecords) SetupErrorResponse(err error) {
+
+	m.responses = append(m.responses, &mockKinesisPutRecordsResponse{
+		Err:    err,
+		Output: nil,
+	})
+}
+
+func (m *mockKinesisPutRecords) PutRecords(input *kinesis.PutRecordsInput) (*kinesis.PutRecordsOutput, error) {
+
+	var reqNum = len(m.requests)
+	if reqNum > len(m.responses) {
+		return nil, fmt.Errorf("Response for request %+v not setup", reqNum)
+	}
+
+	m.requests = append(m.requests, input)
+
+	var resp = m.responses[reqNum]
+	return resp.Output, resp.Err
+}
+
+func (m *mockKinesisPutRecords) AssertRequests(
+	assert *assert.Assertions,
+	expected []*kinesis.PutRecordsInput,
+) {
```

Comment:
Not overly exciting assertions for this PR, but this will be helpful for testing batching behavior in future PRs.  The current test coverage is lacking, so if I can set a good precedent.  :)